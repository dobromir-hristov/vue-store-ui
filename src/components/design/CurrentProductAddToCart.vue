<template>
  <v-button
    :disabled="!canAddToCart"
    type="primary"
    @click="addToCartHandler"
  >
    Add to Cart
  </v-button>
</template>

<script>/**
 * @module CurrentProductAddToCart
 */
import { CurrentProductAddToCartMixin } from '@/mixins/core/CurrentProductAddToCartMixin'

export default {
  name: 'CurrentProductAddToCart',
  mixins: [CurrentProductAddToCartMixin]
}
</script>
