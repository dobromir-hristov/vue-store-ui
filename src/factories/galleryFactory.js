export const galleryFactory = () => ([{
  variation_id: 'color__blue',
  src: 'https://demo.storefrontcloud.io/img/600/744/resize/w/s/ws12-blue_main.jpg'
}, {
  variation_id: 'color__orange',
  src: 'https://demo.storefrontcloud.io/img/600/744/resize/w/s/ws12-orange_main.jpg'
}, {
  variation_id: 'color__purple',
  src: 'https://demo.storefrontcloud.io/img/600/744/resize/w/s/ws12-purple_main.jpg'
}])
