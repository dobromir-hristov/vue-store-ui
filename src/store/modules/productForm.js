import { ProductForm } from '@/models/ProductForm'

const PREPARE_PRODUCT_FORM = 'PREPARE_PRODUCT_FORM'

const state = {
  /** @type ProductForm */
  productForm: null
}

const getters = {
  productForm: state => state.productForm,
  /**
   * Returns the SKU of a product based on the selected variations
   * @param state
   * @return {string}
   */
  formSKU (state) {
    if (!state.productForm) return ''
    return state.productForm.sku + '-' + Object.values(state.productForm.variations).join('-')
  }
}

const mutations = {
  [PREPARE_PRODUCT_FORM] (state, product) {
    state.productForm = new ProductForm(product)
  }
}

const actions = {
  prepareForm ({ commit }, product) {
    commit(PREPARE_PRODUCT_FORM, product)
  }
}

export default {
  state,
  getters,
  mutations,
  actions,
  namespaced: true
}
