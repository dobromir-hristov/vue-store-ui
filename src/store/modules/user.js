/* State is a factory function so it does not pollute in tests */
import { basicSetter } from '../StoreUtils'
import { AuthService } from '../../services/api/AuthService'
import { StorageService } from '../../services/StorageService'

const IS_LOGGING_IN = 'IS_LOGGING_IN'
const IS_FETCHING_USER = 'IS_FETCHING_USER'

const SET_USER = 'SET_USER'
const SET_TOKEN = 'SET_TOKEN'

const state = () => ({
  // loaders
  loggingIn: false,
  fetchingUser: false,
  // state
  user: { id: null },
  token: StorageService.get('token')
})

const getters = {
  userToken: state => state.token,
  isLoggedIn: (state) => !!state.user.id,
  userId: state => state.user.id
}

const mutations = {
  // loading
  [IS_FETCHING_USER]: basicSetter('fetchingUser'),
  [IS_LOGGING_IN]: basicSetter('loggingIn'),
  // state
  [SET_USER]: basicSetter('user'),
  [SET_TOKEN]: basicSetter('token')
}

const actions = {
  async loginUser ({ state, commit, dispatch }, credentials) {
    commit(IS_LOGGING_IN, true)
    try {
      const token = await AuthService.loginUser(credentials)
      commit(SET_TOKEN, token)
      dispatch('fetchUser')
    } finally {
      commit(IS_LOGGING_IN, false)
    }
  },
  async fetchUser ({ state, commit }) {
    commit(IS_FETCHING_USER, true)
    try {
      const user = await AuthService.fetchUser(state.token)
      commit(SET_USER, user)
    } finally {
      // I am not catching the error so we can handle it where necessary
      commit(IS_FETCHING_USER, false)
    }
  }
}

export default {
  state,
  getters,
  mutations,
  actions,
  namespaced: true
}
