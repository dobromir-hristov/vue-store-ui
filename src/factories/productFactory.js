import { galleryFactory } from '@/factories/galleryFactory'

export const productFactory = () => ({
  id: 'some-random-id',
  name: 'Radiant Tee',
  sku: 'radiant-tee',
  price: 22.50,
  currency: 'USD',
  description: `<p>So light and comfy, you'll love the Radiant Tee's organic fabric, feel, performance and style. You may never want to stop moving in this shirt.</p>
<p>• Salmon soft scoop neck tee.<br>• Athletic, semi-form fit.<br>• Flat seams prevent chafing.<br>• 67% Organic Cotton / 28% Hemp / 5% Spandex.</p>`,
  attributes: [
    { type: 'material', values: ['Hemp', 'Organic Cotton', 'Spandex'] },
    { type: 'style', values: ['Tee'] },
    { type: 'climate', values: ['Indoor', 'Warm'] },
    { type: 'pattern', values: ['Solid'] }
  ],
  variations: [{
    id: 'color',
    type: 'color',
    name: 'Color',
    price: 5,
    types: [
      { id: 'blue', name: 'Blue', value: 'blue' },
      { id: 'purple', name: 'Purple', value: 'purple' },
      { id: 'orange', name: 'Orange', value: 'orange' }
    ]
  }, {
    id: 'size',
    type: 'size',
    name: 'Size',
    price: 10,
    types: [
      { id: 'm', name: 'Medium', value: 'm' },
      { id: 'l', name: 'Large', value: 'l' },
      { id: 'xl', name: 'XLarge', value: 'xl' }
    ]
  }, {
    id: 'material',
    type: 'material',
    name: 'Material',
    price: 0,
    types: [
      { id: 'cotton', name: 'Cotton', value: 'cotton' },
      { id: 'cot50poli50', name: 'Cotton 50%, Polyester 50%', value: 'cot50poli50' }
    ]
  }
  ],
  categories: [
    { id: '2', name: 'Default Category' },
    { id: '3', name: 'Women' },
    { id: '4', name: 'Tops' },
    { id: '5', name: 'Tees' }
  ],
  gallery: galleryFactory()
})
