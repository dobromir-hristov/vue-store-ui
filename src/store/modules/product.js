import { Product } from '@/models/Product'
import { ProductService } from '@/services/api/ProductService'
import productForm from './productForm'

const SET_PRODUCT = 'SET_PRODUCT'

const state = () => ({
  /** @type Product */
  product: null
})

const getters = {
  productLoaded: state => !!state.product,
  product: state => state.product,
  currency: state => state.product.currency
}

const mutations = {
  [SET_PRODUCT] (state, product) {
    state.product = new Product(product)
  }
}

const actions = {
  async fetchProduct ({ commit, dispatch }, productId) {
    const product = await ProductService.fetchProduct(productId)
    commit(SET_PRODUCT, product)
    dispatch('productForm/prepareForm', product)
  }
}

export default {
  state,
  getters,
  mutations,
  actions,
  modules: { productForm },
  namespaced: true
}
