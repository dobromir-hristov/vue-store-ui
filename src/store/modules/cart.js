import { basicSetter } from '@/store/StoreUtils'

export const REMOVE_ITEM = 'REMOVE_ITEM'
export const SET_ITEMS = 'SET_ITEMS'
export const ADD_ITEM = 'ADD_ITEM'

const state = {
  items: []
}

const getters = {
  itemsInCart (state) {
    return state.items
  },
  itemsInCartCount (state, getters) {
    return getters.itemsInCart.length
  }
}

const mutations = {
  [REMOVE_ITEM] (state, productId) {
    const index = state.items.findIndex(i => i.id === productId)
    if (index !== -1) {
      state.items.splice(index, 1)
    }
  },
  [SET_ITEMS]: basicSetter('items'),
  /**
   * Adds a product to the items list
   * @param {ProductForm} product
   */
  [ADD_ITEM] (state, product) {
    const index = state.items.findIndex(p => p.sku === product.sku)
    if (index !== -1) {
      state.items[index].quantity += product.quantity
    } else {
      state.items.push(product)
    }
  }
}

const actions = {
  removeItem ({ commit }, productId) {
    commit(REMOVE_ITEM, productId)
  },
  addItem ({ commit }, product) {
    commit(ADD_ITEM, product)
  }
}

export default {
  state,
  getters,
  mutations,
  actions,
  namespaced: true
}
