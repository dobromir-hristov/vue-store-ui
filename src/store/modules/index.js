import user from './user'
import cart from './cart'
import product from './product'

export default {
  user,
  cart,
  product
}
