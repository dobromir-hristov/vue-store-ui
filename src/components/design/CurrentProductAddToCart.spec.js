import CurrentProductAddToCart from './CurrentProductAddToCart'
import modules from '@/store/modules'
import { productFactory } from '@/factories/productFactory'

/**
 * @param overrides
 * @returns {{config, wrapper: Wrapper<V> | Wrapper<Vue> | *}}
 */
const createWrapper = async (overrides) => {
  const store = modules
  let config = _merge(createComponentMocks({ store }), overrides)
  await config.store.dispatch('product/productForm/prepareForm', productFactory())
  const wrapper = mount(CurrentProductAddToCart, config)
  return {
    config,
    wrapper
  }
}

describe('CurrentProductAddToCart', () => {
  it('renders the CurrentProductAddToCart', async () => {
    const { wrapper } = await createWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('does disables button if quantity is 0', async () => {
    const { wrapper, config } = await createWrapper()
    config.store.commit('product/productForm/SET_QUANTITY', 0)
    expect(wrapper.find('button').attributes('disabled')).toBeTruthy()
  })

  it('adds an item to the cart', async () => {
    const { wrapper, config } = await createWrapper()
    wrapper.find('button').trigger('click')
    expect(config.store.getters['cart/itemsInCartCount']).toBe(1)
  })
})
