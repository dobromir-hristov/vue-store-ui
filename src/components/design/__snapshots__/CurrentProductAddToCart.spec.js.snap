// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentProductAddToCart renders the CurrentProductAddToCart 1`] = `
<button class="v-button v-button--primary" type="primary"><span class="v-button__text">
  Add to Cart
</span></button>
`;
